pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -0,0 +1,173 @@
+        segmentZKMetadata.getSegmentName(), kafkaTopic, serverMetrics);
+    _realtimeSegment.setSegmentMetadata(segmentZKMetadata, schema);
+
+    // Create message decoder
+    final KafkaMessageDecoder messageDecoder = (KafkaMessageDecoder) Class.forName(decoderClassName).newInstance();
+    messageDecoder.init(new HashMap<String, String>(), _schema, kafkaTopic);
+
+    // Create field extractor
+    final PlainFieldExtractor fieldExtractor = (PlainFieldExtractor) FieldExtractorFactory.getPlainFieldExtractor(schema);
+
+    // Start indexing thread
+    _indexingThread = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          // Create Kafka consumer
+          SimpleConsumerWrapper consumerWrapper =
+              SimpleConsumerWrapper.forPartitionConsumption(bootstrapNodes, kafkaTopic, kafkaPartitionId);
+
+          // TODO Check for limit conditions (eg. full, stop consuming)
+          boolean notFull = true;
+          long currentOffset = _segmentZKMetadata.getStartOffset();
+
+          while (notFull) {
+            // Get a batch of messages from Kafka
+            Iterable<MessageAndOffset> messagesAndOffsets =
+                consumerWrapper.fetchMessages(currentOffset, Long.MAX_VALUE, 1000);
+
+            // Index each message
+            int batchSize = 0;
+            for (MessageAndOffset messageAndOffset : messagesAndOffsets) {
+              byte[] array = messageAndOffset.message().payload().array();
+              int offset = messageAndOffset.message().payload().arrayOffset();
+              int length = messageAndOffset.message().payloadSize();
+              GenericRow row = messageDecoder.decode(array, offset, length);
+
+              if (row != null) {
+                row = fieldExtractor.transform(row);
+                notFull = _realtimeSegment.index(row);
+                batchSize++;
+              } else {
+                System.out.println("Dropped null row?");
+              }
+              currentOffset = messageAndOffset.nextOffset();
+            }
+
+            if (batchSize != 0) {
+              System.out.println("Indexed " + batchSize + " messages from partition " + kafkaPartitionId + " offset " + currentOffset);
+            } else {
+              Uninterruptibles.sleepUninterruptibly(100, TimeUnit.MILLISECONDS);
+            }
+          }
+        } catch (Exception e) {
+          LOGGER.error("Caught exception while indexing events", e);
+        }
+      }
+    }, "Realtime indexing thread for " + _segmentZKMetadata.getSegmentName());
+    _indexingThread.start();
+  }
+
+  @Override
+  public IndexSegment getSegment() {
+    return _realtimeSegment;
+  }
+
+  @Override
+  public String getSegmentName() {
+    return _realtimeSegment.getSegmentName();
+  }
+
+  @Override
+  public void destroy() {
+    // TODO
+  }
+
+  public void goOnlineFromConsuming() {
+    // Download segment
```

Comment:
update the comments here to reflect our discussion. You cannot always download and replace the segment. The code in validation branch does it right.